src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -68,5 +96,287 @@ protected static SymbolDisplayPart NewLine()
+        {
+            var argumentCount = arguments.Count;
+            var argToParamMap = ArrayBuilder<int>.GetInstance(argumentCount, -1);
+            var parameters = method.Parameters;
+            var parameterCount = parameters.Length;
+            var currentParameterIndex = 0;
+            var seenOutOfPositionArgument = false;
+            var inParams = false;
+
+            for (var argumentIndex = 0; argumentIndex < argumentCount; argumentIndex++)
+            {
+                if (argumentIndex >= parameterCount && !inParams)
+                {
+                    argToParamMap.Free();
+                    return null;
+                }
+
+                var argument = arguments[argumentIndex];
+                if (HasName(argument, out var name))
+                {
+                    var namedParameterIndex = parameters.IndexOf(p => p.Name == name);
+                    if (namedParameterIndex < 0)
+                    {
+                        argToParamMap.Free();
+                        return null;
+                    }
+
+                    if (namedParameterIndex != currentParameterIndex)
+                    {
+                        seenOutOfPositionArgument = true;
+                    }
+
+                    AddArgToParamMapping(argumentIndex, namedParameterIndex);
+                    IncrementParameterIndexIfNeeded();
+                }
+                else if (IsEmptyArgument(argument.Expression))
+                {
+                    if (!seenOutOfPositionArgument)
+                    {
+                        // We count the empty argument as a used position
+                        AddArgToParamMapping(argumentIndex, currentParameterIndex);
+                        IncrementParameterIndexIfNeeded();
+                    }
+                }
+                else if (seenOutOfPositionArgument)
+                {
+                    // Unnamed arguments are not allowed after an out-of-position argument
+                    argToParamMap.Free();
+                    return null;
+                }
+                else
+                {
+                    AddArgToParamMapping(argumentIndex, currentParameterIndex);
+                    IncrementParameterIndexIfNeeded();
+                }
+            }
+
+            return argToParamMap;
+
+            void IncrementParameterIndexIfNeeded()
+            {
+                if (!seenOutOfPositionArgument && !inParams)
+                {
+                    currentParameterIndex++;
+                }
+            }
+
+            void AddArgToParamMapping(int argumentIndex, int parameterIndex)
+            {
+                Debug.Assert(parameterIndex >= 0);
+                Debug.Assert(parameterIndex < parameterCount);
+
+                if (parameters[parameterIndex].IsParams)
+                {
+                    inParams = true;
+                }
+
+                argToParamMap[argumentIndex] = parameterIndex;
+            }
+        }
```

Comment:
i think it would be good to extract this out to a dedicate type.  something like LightweightOverloadResolution or something like that.